### ISSUE TITLE
New Approach to Move Cache dir to external SD

### ISSUE THREAD

**Issue by marco-dev** [cgeo/cgeo#4957]
Hi together,

I would like to try a new approach moving cache dir to external sd like Locus does.
Please vote for or against a new approach.

Choices for user:
- /storage/emulated/0/.cgeo (default)
- /storage/extSdCard/Android/data/cgeo.geocaching/.cgeo (will be removed on uninstall)

The files should be moved by c:geo with progress dialog.
We already have a function to remove garbage from cache directory.
I suggest to remove garbage before moving files.

Open questions, please give your feedback:
- Should I do another approach?
- If the "moving files" dialog is cancel by back button or by sending the app to background, what should be done?
- Should we remove garbage before moving cache directory?
- Any other questions?

**Comment by Cappa-d** [comment]
As I ran into diskspace issues last night I was close to get rid of c:geo and move to locus last night... so glad this came up again.

I am eager to see a solution to have c:geo on the real extSDCard! too much issues here and I think if someone can identify the difference between emulated and extSDCard we are not talking about a "normal" user who must be protected because otherwise the app gets blamed.

Moving files: would it be possible to trigger the following cascade:

a) 1:1 copy from internal to extSDCard
b) move 100% trigger deletion on emulated (would avoid an issue with "lost" data due to interrupted dialog)
c) deletion 100% triggers final setting changes (path for internal storage)


Garbage removal:

May add this as an internal cleanup service / menu item which could be triggered at any time from the settings menu. Would prefer to separate this from the moving part.

**Comment by marco-dev** [comment]
@Cappa-d 
Garbage removal is already a function in settings, besides I did never try it nor know if this removes all or only garbage data. Have to read FAQ again. :-)

**Comment by Lineflyer** [comment]
> Should I do another approach?

Yes I would highly appreciate that. Thank you very much for coming back to this.

First my thought was, that the topic will get less important as the amount of memory in the device is also increasing with new models. But current user reports show, that also the amount of cache data is rising (more caches, more pictures, etc.) and a lot of users complain about too less space on their internal disc.

> If the "moving files" dialog is cancel by back button or by sending the app to background, what should be done?

Do we need a "Cancel"?

> Should we remove garbage before moving cache directory?

Might remove the amount of unneccesary data to be moved, so yes. And we already have such a function which does only need to be called.

> Any other questions?

I think those two options 
    /storage/emulated/0/.cgeo (default)
    /storage/extSdCard/Android/data/cgeo.geocaching/.cgeo (will be removed on uninstall)
are the only feasible method to keep the complexity as minimal as possible.
If I look into the other issues we had concerning this topic, I learned that all other scenarios would require extensive checks of writability and highly depend on Android verisons, etc.etc.

One remark:
Due to the fact that  /storage/extSdCard/Android/data/cgeo.geocaching/ is not accesible by the user and also removed upon uninstall we should in parallel modify the backup/restore function to target a separate folder (at the moment its targeting the very same folder).

**Comment by SammysHP** [comment]
> Do we need a "Cancel"?

No, but the process might be interrupted.

> Due to the fact that /storage/extSdCard/Android/data/cgeo.geocaching/ is not accesible by the user and also removed upon uninstall we should in parallel modify the backup/restore function to target a separate folder (at the moment its targeting the very same folder).

But we cannot write to other directories. AFAIK there are some ways to bypass this limitation, but I've never looked into detail for this.

> Any other questions?

- Parallel with the option for database on external storage?
- What if the user removes the SD card and starts c:geo? (And what while it is running?)
- Will it then ask to fall back to internal memory? What if the DB is stored in /data, so only images etc. are missing?
- What if the user has the DB stored in /sdcard (internal) and uses the internal memory as well as the real SD card (as a consequence of the previous question)? Then there are two DB files, how to merge them?

**Comment by rsudev** [comment]
My impression from the previous discussions was, that we agreed to decouple database storage and backup from geocache-related file storage.
This would prevent a number of possible issues, and the storage benefit of moving the cache database to the external sd card is relatively low IMO.
So my suggestion is, to leave the handling for the database as it is and allow moving the file storage to this defined location.
Of course it can be a challenge, to find a wording which a user without technical background can understand and also is true for the variety of platforms we support (2.3 -> 5.1).
It might be a good idea to take this opportunity to also move the default folder from .cgeo to cgeo/cachedata and perhaps have also cgeo/database and cgeo/backup, but should probably rather be a separate item.
So database could be stored in:
* app storage (default)
* sdcard/internal

and cache files in:
* sdcard/internal (default)
* external

and backup will always go to:
* sdcard/internal

Given this, in my opinion it is not necessary to put too much emphasis on the transactional nature of the move process. I would say, change the setting, then start moving (this could even be optional). If it is interrupted, then you would have to re-download a couple of images, not a big deal IMO.
When setting back, the move back can be started, and if we do not forcefully clean the target, you can re-fuse your stored files, if so wished. Additionally clean out options for both areas can be offered, as long it is ensured, that the (possibly external) database and the backup is not affected.

**Comment by Lineflyer** [comment]
@SammysHP
Regarding backup/restore I was not suggesting to write it to external SD but similar to the existing approach to store it always on the internal SD maybe just in /
Mainly just make it independent from the general storage location


DB on external should no longer be a separate but combined with the general setting. Of course all possible  migration cases have to be handled. This would make the last two questions obsolete.

For removing SD while data is stored there we already had a good approach in another thread AFAIR.
I think this was detecting it and asking to replace or startover with a new DB on internal. This should also cover the case of the SD being formatted, etc.

**Comment by Lineflyer** [comment]
@rsudev
While those ideas are good, I would prefer to keep the first step as simple as possible. Just handle DB storage, backup files and cached data.

Once this is working we can work on the details of all the other folders/files we are generating.

**Comment by mucek4** [comment]
Could we made backup to online serves? Gmail, dropbox, seafile, ...

**Comment by SammysHP** [comment]
@rsudev 
From what I've heard one problem is that the database file grows too much, so it must be possible to move it the the real SD card.

@mucek4 
That's a different issue.

**Comment by Lineflyer** [comment]
@SammysHP
Thats why I suggested to just combine that settings of DB and cache folder.  Users need more space, they probably dont care which data is behind it

**Comment by rsudev** [comment]
If we make backup target an independent location (like /sdcard/cgeo/backup) always, I am fine with whatever approach, even though I do not buy the database size argument.

**Comment by rsudev** [comment]
I have to modify my last statement slightly: I would like to retain the possibility to keep the database in appstorage, regardless of any setting for the cache directory.

**Comment by Lineflyer** [comment]
Let me try to summarize before this thread will get too long:

- We seem to agree that another approach for implementation is appreciated

- The two choices for the cache folder are OK as mentioned:
    /storage/emulated/0/.cgeo (default as it is now)
    /storage/extSdCard/Android/data/cgeo.geocaching/.cgeo (will be removed on uninstall)

- The folder for backup/restore needs to be a separate (new) one to avoid the backup being stored in a non user accesible folder.
This could even be implemented in an early step (right now) to have clear conditions for a future version implementing the external storage.

- We should not put too much effort in moving. Just try it if the user whiches, otherwise just switch to the new location and maybe delete the old folder.

- Regarding the DB storage there have been different opinions, but rethinking different scenarios (e.g. as mentioned by @SammysHP) it would be best to remove that "DB on SD" option completely (at least in the first step) and always store it at the internal system dir.
This has the big advantage, that even if the SD is removed or the content deleted, the user will have all his lists but only loose some images, etc. The DB needs to be safe as it is mandatory to run c:geo.
So we need to implement a one time fallback by moving the DB in case it is on "external" when starting c:geo after the upgrade to the version. 
This could even be implemented in an early step (right now) to have clear conditions for a future version implementing the external storage.

- On startup of c:geo and if external storage is active, it should be checked whether it is available. If not, the user should be made aware of this and get the choice to close c:geo (and we keep the link to the external) or run anyway (and we fall back to the internal folder).

I hope I did not forget an important thing.

**Comment by mucek4** [comment]
Only think is. On my HTC the internal space is /storage/emulated/0/ and SD card is on /storage/ext_sd/

**Comment by Lineflyer** [comment]
> Only think is. On my HTC the internal space is /storage/emulated/0/ and SD card is on /storage/ext_sd/

Should be no problem at all:
Paths for both options are delivered by the android system as fas as I understood that.

**Comment by SammysHP** [comment]
> Paths for both options are delivered by the android system as fas as I understood that.

Nope.

> it would be best to remove that "DB on SD" option completely

I've seen DB files with more than 200 MB and users complaining about not enough free storage space.

**Comment by rsudev** [comment]
Making backup address a new and different directory on sdcard/internal as a preparation step sounds like a very good idea. Issue created: #4963

**Comment by marco-dev** [comment]
We should not mix up database file with cache directory. This should go into another issue.

So one step going before this can be implemented is to separate cache directory path from db file path.

**Comment by Lineflyer** [comment]
So now we started two new issue #4963 and #4964 to do necessary preparation.
I would suggest to start these activities right **after** next feature release (respectively after copying `master` to `release`) which is planned for hopefully next week (refer to #4953).

**Comment by marco-dev** [comment]
Yes. Just as a feedback: I already prepared the current master on my local repository and downloaded all the new APIs. :-)

**Comment by marco-dev** [comment]
Should we migrate the cache dir to visible "cgeo" subdir (without leading dot)?
Do we have problems to make it visible because the images will be found by gallery app?

**Comment by SammysHP** [comment]
Simply put a `.nomedia` file in that directory, this excludes all media from the indexer.

**Comment by marco-dev** [comment]
Because no one answered my question here:

**Should we migrate the cache / backup / db dir to visible "cgeo" subdir (without leading dot)?**

**Comment by Lineflyer** [comment]
Without dot... as our target design might be, that those files stored there should be accessible by the user without a problem lateron.

**Comment by Lineflyer** [comment]
I added a section about GPX import/export to the Wiki "Target structure" page.
Hope this is correct and the migration proposal will work.

**Comment by Lineflyer** [comment]
@marco-dev 
Maybe we should really "meet" on IRC with some people who want to help defining the whole thing.
Any timing proposal?

**Comment by SammysHP** [comment]
Another idea: WebRTC

**Comment by marco-dev** [comment]
At work 7:20

**Comment by madankumarbaskaran** [comment]
Hi @marco-dev. Please let me know if you like to share some work on storage issue. Happy to work with you.

**Comment by Cappa-d** [comment]
any news on this topic? Despite having upgraded my mobile last year around this time and having 4 x the available internal memory now. I am already running into the same problem and so I am highly interested to  test any changes here..

Please state here if there is anything testing needed etc. unfortunately I cannot help with the coding part :(

**Comment by Lineflyer** [comment]
@Cappa-d 
No, there is no active developer right now working on this.
If you know someone who wants to go ahead, please feel free to contact us,

**Comment by Cappa-d** [comment]
Dang... on friday night the worst case happened and the fact that the C:geo data was not on my external SD card caused me to loose everything now.. Handy died and now I have lost what was stored only in C:geo... sigh

After this mess I would love to help here but no skills in this area unfortunately :(

**Comment by valentijnscholten** [comment]
I think there's also a use case to keep the database on the internal memory to keep it fast, but move the cache data (and backups?) to external sd card.
By external I mean a physical sd card that is insert into the phone, not the emulated sdcard which is basically internal phone memory.

**Comment by pstorch** [comment]
There are two wiki docs regarding this issue (and the subissues):
1. https://github.com/cgeo/cgeo/wiki/Ideas-and-Migration-steps
1. https://github.com/cgeo/cgeo/wiki/Target-disk-usage-structure

It seems that the 2. is the newer one and replaces the 1., right?
@Lineflyer if yes, should we remove the older one?

**Comment by pstorch** [comment]
I've updated the https://github.com/cgeo/cgeo/wiki/Target-disk-usage-structure. The field notes and the local spoiler images should also be moved under `<externalStorageDir>/cgeo`, right?

**Comment by Lineflyer** [comment]
I assume so. AFAIR the `<externalStorageDir>` is the base path of where we currently also store in `/.cgeo/` right?

**Comment by pstorch** [comment]
@Lineflyer yes it is

**Comment by pstorch** [comment]
I stumbled upon the `.corrupted` database. Which backs up the current database if there is an error opening it. At the moment this is created in the same directory as the database file was. In case of the internal directory this file is not easily accessible. I think we should create this file in the new `backup` directory, too. This way a user can find it and maybe send it to us for inspection.

**Comment by pstorch** [comment]
**Small Update**
At the moment I'm concentrating all logic about local directories into the `LocalStorage` class and move everything out (mostly to `FileUtils`) which is not specific to these directories.

**Regarding the cache dir**
While reading more about the [(external) storage options on Android](https://developer.android.com/guide/topics/data/data-storage.html) I think we should use [Context.html#getExternalCacheDir()](https://developer.android.com/reference/android/content/Context.html#getExternalCacheDir()) as the default cache dir and on API Level 19+ we can offer to choose from [Context.html#getExternalCacheDirs()](https://developer.android.com/reference/android/content/Context.html#getExternalCacheDirs()) (plural) and store that in the Settings. But if the chosen directory isn't available (e.g. not mounted) we should fallback to the default [Context.html#getExternalCacheDir()](https://developer.android.com/reference/android/content/Context.html#getExternalCacheDir()). Which is usually the emulated one.

There is also an [internal cache dir](https://developer.android.com/reference/android/content/Context.html#getCacheDir()) but this should be limited to 1MB. With the cached images we are above that very quickly and should avoid it from the beginning.

**Comment by Lineflyer** [comment]
AFAIR what you describe is also what we discussed last time:
The cache dirs should end up in a dedicated directory, which is removed when the app is uninstalled (this is the case for both options you describve). 
The only difficulty is, to move all files which might be needed to be accessed by the user out of this directory in a first migration step (e.g. backup, etc. as mentioned above or in Wiki).

One remark regarding possible timing:
Let us first finish the minsdk change (wait until next feature release is out) before starting any change.

**Comment by pstorch** [comment]
Regarding timing: this is my understanding, too. I'll need some more time anyway.

**Comment by pstorch** [comment]
We could use a private external dir for the DB with [Context.html#getExternalFilesDir(java.lang.String)](https://developer.android.com/reference/android/content/Context.html#getExternalFilesDir(java.lang.String)) but then the DB gets deleted when uninstalling c:geo and "There is no security enforced with these files".  So the user or other apps might still be able to access the DB. I think it depends on if the external directory is an emulated storage device or a real external sd card.

**Comment by pstorch** [comment]
Continuing here:

Just tried with `Context.html#getExternalFilesDir(null) + "/databases"` then on my device the database file will be stored in `/sdcard/Android/data/cgeo.geocaching/files/databases/`. It is more hidden than `/sdcard/cgeo/databases` but still accessible, if the user knows where. I think I like it, let's go for this option.

The external cache is also going nearby `/sdcard/Android/data/cgeo.geocaching/cache/`.
BTW: if we use this "official" cache directory then the user can clean it up via the Android "App info" page -> storage -> clear cache. I think this is good.

**Comment by Lineflyer** [comment]
Thanks @pstorch for trying this.

`/sdcard/Android/data/cgeo.geocaching/files/databases` 
Yes, it is indeed quite hidden`, but if this is the target for the database itself, the user won't need access to it (and maybe should not have). How are these files handled by the system? Is this removed on uninstall?

`/sdcard/Android/data/cgeo.geocaching/cache/`
When you say that this is the official "cache" directory, then I think it is not wise to use this for this kind of data. Speaking about "cache" and especially the "Clear cache" function I would not expect that it deletes important data....and the data which we save there is kind of important but not just some temporary stuff.
Furthermore there are a lot of "cleanup" tools/apps available which simply wipe all these cache directories regularly. This would mean the user looses all offline data for his saved caches.

**Comment by pstorch** [comment]
The full `/sdcard/Android/data/cgeo.geocaching` directory gets deleted on uninstall.

Regarding the "cache": technically all these files can be restored by their online version, so there is nothing "lost". But I get your point. Usually the user wants to keep these files for offline caching and would be surprised if they are suddenly gone. Then we should put the cache also under `/sdcard/Android/data/cgeo.geocaching/files/cache`, so that it is only deleted on uninstall or by our own housekeeping routine.

**Comment by Lineflyer** [comment]
> Then we should put the cache also under `/sdcard/Android/data/cgeo.geocaching/files/cache`, so that it is only deleted on uninstall or by our own housekeeping routine.

Yes, that is better.
One small remark: I would go for the term `GeocacheData` rather than `cache` for the very same reason (in case of manual "cleanup" tries by the user).

**Comment by pstorch** [comment]
@Lineflyer `GeocacheData`, yes that's a better name

**Comment by Lineflyer** [comment]
Updated https://github.com/cgeo/cgeo/wiki/Target-disk-usage-structure accordingly.

**Comment by Lineflyer** [comment]
Another small remark while reading the Wiki:
We should probably rename `GeocachePhotos` during migration to `GeocacheSpoilers` as this is the purpose of this folder.
Or is the name `GeocachePhotos` predefined by Garmin (or whereever this spoiler sync function is derived from).

**Comment by Lineflyer** [comment]
And last but not least:
The Wiki seems to miss the folder, where we store photos taken with c:geo (for adding as a log picture). As I don't know the current implementation someone else needs to find out and add it to the wiki.

**Comment by pstorch** [comment]
Regarding `GeocachePhotos`: #3438
And when I see this commit https://github.com/cgeo/cgeo/commit/d48b4a67639202e6fdc8599740ad68619034a2b4 it was called `GeocacheSpoilers` before.

I'll check where we put the photos taken with c:geo.

**Comment by pstorch** [comment]
Logphotos taken with cgeo are stored in `/sdcard/Pictures/cgeo`. This is a result of `Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_PICTURES) + "/cgeo"`.
I'll move this logic into the `LocalStorage` class as well, to have everything local directories related in one place.

**Comment by pstorch** [comment]
I've updated the https://github.com/cgeo/cgeo/wiki/Target-disk-usage-structure page once more.
Added a section about the general Android directories at the top and referenced them afterwards. Please review.
I'm not sure if we should move the log Pictures.

**Comment by Lineflyer** [comment]
A good idea, however I do not see any of the references to the android directories in the chapters below (or at least not with the same name (maybe you can number the different dirs on top and then refer to them by the number to make it more clear).

**Comment by pstorch** [comment]
Good idea, I've numbered the standard directories. But now we have to look up into the reference section to know which directory is meant.

**Comment by pstorch** [comment]
I would like to test the storage on a real external sdcard (referenced by <5>), but I don't have such a device. Mine has only emulated sdcard. I try to figure out how to setup such a device on the emulator.

**Comment by Lineflyer** [comment]
If you are not succesful with an emulator either I can offer to make some tests with my old S4, else I can buy you a test device (this task is well worth it IMHO).

**Comment by pstorch** [comment]
I'm now extending the Settings screen. The user should be able to move the `GeocacheData` directory between an "emulated" and a "real" extSdCard. I thought about presenting a list dialog with all available external storages (emulated and real). What if the device doesn't have "real" external storages available? Should we still show the setting and a list with only one entry or should I hide this option in this case?

**Comment by Lineflyer** [comment]
I took the time and installed the corresponding the debug build based on this PR from CI.
Before doing that I installed a debug version based on an older stable PR (http://ci.cgeo.org/job/cgeo%20pull%20request/786/) to be able to do the migration step.

Before installing the debug version based on this PR I did the following:
- Enter login data
- Changed setting to store log pictures and static maps offline
- Store 35 caches distributed over two lists
- Export a GPX file
- Export two field notes
- Take a log picture photo with c:geo
- Create a database backup

Afterwards I created a backup copy of all the data in the file system I produced while doing this.
Then I installed the new debug version and started it once waiting for login to be finished.

Now I analyzed the new status of the app and files as follows:
- Lists are still complete, so database move was successful. I do now see the DB stored in `/storage/emulated/0/Android/data/cgeo.geocaching/files/databases/` while it was in non accesible folder before
- Additional data for caches has correctly been moved from `/.cgeo/*` to `/storage/emulated/0/Android/data/cgeo.geocaching/files/GeoacheData` and access to it is working (tested by opening static maps and log picture tab for a saved cache)
- The backup previously created and saved into `/.cgeo/cgeo.sqlite` has **NOT** been migrated. The new version tells me that no backup is available and the folder `/storage/emulated/0/Android/data/cgeo.geocaching/backup` does not even exist. The existing `/.cgeo/cgeo.sqlite` is no longer there. Backup is lost.
- The exported GPX files have been migrated from the default dir `/storage/emulated/0/gpx/` to `storage/emulated/0/cgeo/gpx`
- The exported field notes have been migrated from the default dir `/storage/emulated/0/field-notes/` to `storage/emulated/0/cgeo/field-notes`
- The existing log picture in `/storage/emulated/0/Pictures/cgeo` has not been migrated and is still stored at its original location. 

Conclusion:

Migration step | Result|  Comment
--------- | ----------- | -----------
Database | PASSED | Small remark: The dir could be named `database` instead of `databases`
Additional cache data | PASSED |
Backup | FAILED | Backup is lost
Exported GPX | PASSED |
Exported Field Notes | PASSED |
Log Pictures Taken | INCONCLUSIVE | Was this planned to be migrated?

**Comment by Lineflyer** [comment]
[22:04] <Lineflyer> Regarding the database: i think you mixed up something here. I try to explain:
[22:04] <pstorch> Regarding DB: the switch moves between /data/data/cgeo.geocaching/databases and /Android/data/cgeo.geocaching/files/databases
[22:04] <Lineflyer> data/data ?
[22:05] <pstorch> yes, that is the internal storage, where you don't have access without root
[22:05] <Lineflyer> Oh something is mixed up here I guess. Let me try to explain
[22:05] <pstorch> the default directory where the app can store its data
[22:05] <pstorch> ok
[22:07] <Lineflyer> I found the database now after enabling "DatabaseonSD". it seems to be on the real external SD in storage/extSDcard/Android/data/cgeo.geocaching/files/databases and there are two files named data and data-journal (which I never seen before). I assume thats the real active DB??
[22:07] <Lineflyer> Might it be that the sqlite file in /storage/emulated/0/Android/data/cgeo.geocaching/databases/ is in reality themissing backup?
[22:08] <Lineflyer> In that case database handling is perfectly fine just backup has the problem.
[22:09] <pstorch> active db is called "data" and backup is "cgeo.sqlite". The data.journal is some extra from sqlite.
[22:09] <pstorch> I'm afk for a moment
[22:09] <Lineflyer> so.... /storage/emulated/0/Android/data/cgeo.geoaching/files/databases/cgeo.sqlite is the backup!
[22:10] <Lineflyer> It should not be there as this folder is removed on uninstall. It should be in /storage/emulated/0/cgeo/backup/
[22:10] <Lineflyer> Rest seems fine then if you confirm so far.

**Comment by pstorch** [comment]
Summary of open issues:
- move of GeocacheData between `emulated` and `real` extSdCard not working
- move of sqlite backup file in wrong directory, should go to public external storage: `/storage/emulated/0/cgeo/backup/`
- after migration empty legacy directories should be deleted
- GeocacheData Settings should be moved below the DBonSD Setting

**Comment by Lineflyer** [comment]
I tested the updated PR once again with following result:


Migration step | Result|  Comment
--------- | ----------- | -----------
Database | PASSED | 
Additional cache data | PASSED |
Backup | PASSED |  
Exported GPX | PASSED |
Exported Field Notes | PASSED |

For moving the `GeocacheData`from internal to external SD it is however still not working. This time is does not crash anymore but the existing files are not moved.
Once after having selected the external storage, it stores new GeocacheData without a problem there.

Here is the log:
```
21:45:56.623 Warning cgeo 17991  [RxCachedThreadScheduler-6] Couldn't rename /storage/emulated/0/Android/data/cgeo.geocaching/files/GeocacheData/shared to /storage/extSdCard/Android/data/cgeo.geocaching/files/GeocacheData/shared
21:45:56.623 Warning cgeo 17991  [RxCachedThreadScheduler-6] Couldn't rename /storage/emulated/0/Android/data/cgeo.geocaching/files/GeocacheData/GC352Y3 to /storage/extSdCard/Android/data/cgeo.geocaching/files/GeocacheData/GC352Y3
21:45:56.623 Warning cgeo 17991  [RxCachedThreadScheduler-6] Couldn't rename /storage/emulated/0/Android/data/cgeo.geocaching/files/GeocacheData/GC3EQ4W to /storage/extSdCard/Android/data/cgeo.geocaching/files/GeocacheData/GC3EQ4W
21:45:56.623 Warning cgeo 17991  [RxCachedThreadScheduler-6] Couldn't rename /storage/emulated/0/Android/data/cgeo.geocaching/files/GeocacheData/GC533KT to /storage/extSdCard/Android/data/cgeo.geocaching/files/GeocacheData/GC533KT
21:45:56.623 Warning cgeo 17991  [RxCachedThreadScheduler-6] Couldn't rename /storage/emulated/0/Android/data/cgeo.geocaching/files/GeocacheData/GC40D80 to /storage/extSdCard/Android/data/cgeo.geocaching/files/GeocacheData/GC40D80
21:45:56.623 Warning cgeo 17991  [RxCachedThreadScheduler-6] Couldn't rename /storage/emulated/0/Android/data/cgeo.geocaching/files/GeocacheData/GC3MCKK to /storage/extSdCard/Android/data/cgeo.geocaching/files/GeocacheData/GC3MCKK
21:45:56.623 Warning cgeo 17991  [RxCachedThreadScheduler-6] Couldn't rename /storage/emulated/0/Android/data/cgeo.geocaching/files/GeocacheData/GC6TPD4 to /storage/extSdCard/Android/data/cgeo.geocaching/files/GeocacheData/GC6TPD4
21:45:56.623 Warning cgeo 17991  [RxCachedThreadScheduler-6] Couldn't rename /storage/emulated/0/Android/data/cgeo.geocaching/files/GeocacheData/OC1120A to /storage/extSdCard/Android/data/cgeo.geocaching/files/GeocacheData/OC1120A
21:45:56.623 Warning cgeo 17991  [RxCachedThreadScheduler-6] Couldn't rename /storage/emulated/0/Android/data/cgeo.geocaching/files/GeocacheData/GC150HM to /storage/extSdCard/Android/data/cgeo.geocaching/files/GeocacheData/GC150HM
21:45:56.623 Warning cgeo 17991  [RxCachedThreadScheduler-6] Couldn't rename /storage/emulated/0/Android/data/cgeo.geocaching/files/GeocacheData/GC5HQ4J to /storage/extSdCard/Android/data/cgeo.geocaching/files/GeocacheData/GC5HQ4J
```

**Comment by Lineflyer** [comment]
Not to forget:
Despite the move of `GeocacheData` the implementation looks good to me.
IMHO it could be merged after the move problem has been solved. Other minor problems can be corrected after merge.

**Comment by Lineflyer** [comment]
@pstorch:
After reading https://github.com/cgeo/cgeo/wiki/Target-disk-usage-structure#database once again.
My testing results looks slightly different:
- DB was stored internally with old version (DBonSD not active)
- Also after upgrading to the new model I could not see it in `/storage/emulated/0/Android/data/cgeo.geocaching/files/` which is OK. Nothing seems to have changed.
- After selecting to move it to SD I could see it in `/externalSD/Android/data/cgeo.geocaching/files/databases`

For me this is all OK but seems different to the Wiki description.

**Comment by pstorch** [comment]
Ok, that means that a File rename does not work across mount points. I have to change it to a real copy to the target and delete of the old. 
Regarding the DB: if DBonSD is NOT active then the DB should be on the internal storage referenced by <1>. You don't have access to this directory on normal circumstances. You need to have a root Explorer or via adb USB debugging.

**Comment by pstorch** [comment]
If you want to check the DB from internal storage via adb: https://gist.github.com/15b5d237a65e833a4c40856e2ab7dc38

**Comment by Lineflyer** [comment]
> Regarding the DB: if DBonSD is NOT active then the DB should be on the internal storage referenced by <1>. You don't have access to this directory on normal circumstances. You need to have a root Explorer or via adb USB debugging.

But in the case of DBonSD **is** active, the Wiki should either read `<5>/databased/` instead `<2>/databases` (thats how you implemented it) or the DB should be on `/storage/emulated/0/Android/data/cgeo.geocaching/files/databases/` instead.
We should think about what is best suitable and free of risk: 
Actually if DBonSD is active in current model, the DB is in /storage/emulated/0/.cgeo` (which is the simulated SD). So IMHO the less risky method should be to migrate it to `/storage/emulated/0/Android/data/cgeo.geocaching/files/databases/` in this case as it is on the same mount (no problem with disk space or move procedure).

**Comment by Lineflyer** [comment]
> Ok, that means that a File rename does not work across mount points. I have to change it to a real copy to the target and delete of the old.

See here:
https://developer.android.com/reference/java/io/File.html#renameTo(java.io.File)

**Comment by pstorch** [comment]
Ah, yes. Thanks for looking it up. It's like I suspected. Will change it accordingly.

**Comment by pstorch** [comment]
I've now implemented a fallback to copy+delete the geocache data directories if the rename doesn't work. So we have the fast move, if it is on the same mount point and the longer copy+delete else.
The DBonSD is now kept on the first external dir, which is usually the emulated one (as described in the wiki page).

**Comment by Lineflyer** [comment]
Thanks @pstorch 
Will test it using the debug build tonight or tomorrow. Expect feedback accordingly :)

**Comment by pstorch** [comment]
@Lineflyer fingers crossed ;)

**Comment by Lineflyer** [comment]
I tested the latest status of the PR as of today with following result:

Migration step | Result|  Comment
--------- | ----------- | -----------
Database | PASSED | 
Additional cache data | PASSED |
Backup | PASSED |  
Exported GPX | PASSED |
Exported Field Notes | PASSED |

Also moving the GeocacheData from internal to external SD works now without a problem.

So from testing / usage perspective its good to merge to `master` to allow more intense testing (and some more test users). Can't judge the code review though.

As just discussed on IRC it might be helpful to implement some `Log.i` and `Log.w` during the migration process for more easy debugging later on.

Remaining tests like SDfull, SD removed, Huge amount of caches will be tested in a next step.
For some minor issues I found, I will open dedicated issues once it is merged.

Great work @pstorch
